typing/typecore.mli

```diff
@@ -124,7 +124,7 @@ type error =
   | Multiply_bound_variable of string
   | Orpat_vars of Ident.t * Ident.t list
   | Expr_type_clash of
-      Ctype.Unification_trace.t * type_forcing_context option
+      Ctype.Unification_trace.t * type_forcing_context option * Typedtree.expression_desc option
```

Comment:
Instead of putting the `expression_desc` here, what do you think of always putting it inside the `Diff` constructor? Could it be useful in other situations? cc-ing @Octachron who worked on the unification traces.

Reply:
I was thinking about it but the unification trace only knows about types, not patterns, expressions or literal.

In some sense, this `expression_desc` field is the verso of the `type_forcing_context` info (a `syntactly_forced_type`?), so this location is quite natural. However, it might be worth it  to store only the `Parsetree.constant` info in order to have the same field on both the expression and the pattern error.

Reply:
`Expr_type_clash` is a place where there is a unification error *and* there is a clear (sub)expression that is guilty for the clash. In many other cases, there is a unification failure, but it doesn't really make sense to say that a specific expression is guilty. The way I think of it, `type_forcing_context` and `type_expr` are dual hints for error messages: the former propagates information from the shape of the context around you, the latter propagate information about the shape of the term below you -- used in this particular case for numeric-literal hints. (`Pattern_type_clash` would be a second candidate for this information.)

Reply:
Yes, so this hint could be implemented in `Printtyp.explanation_diff`. But the traces are used on more than expressions (eg. patterns).

Reply:
I think there is consensus that the current approach is a reasonable one -- probably better than overloading unification traces.